index.js

```diff
@@ -218,20 +218,26 @@ Client.prototype._pollConfig = function () {
     }
 
     streamToBuffer(res, (err, buf) => {
-      if (err) return res.destroy(err)
+      if (err) {
+        this.emit('request-error', processConfigErrorResponse(res, buf, err))
+        return
+      }
 
       if (res.statusCode === 200) {
         // 200: New config available (or no config for the given service.name / service.environment)
         const etag = res.headers.etag
         if (etag) this._conf.lastConfigEtag = etag
 
+        let config
         try {
-          this.emit('config', JSON.parse(buf))
-        } catch (e) {
-          res.destroy(e)
+          config = JSON.parse(buf)
+        } catch (parseErr) {
+          this.emit('request-error', processConfigErrorResponse(res, buf, parseErr))
+          return
         }
+        this.emit('config', config)
```

Comment:
> where JSON.parse throws that the client will now emit a config event with config = undefined

It should not. The "return" in the `catch(parseError) { ... }` should handle that.   I just verified that. My mock APM server returned:

```
[2021-02-24T01:25:24.539Z]  INFO: mockapmserver/97351 on pink.local: request (req.remoteAddress=::ffff:127.0.0.1, req.remotePort=64254, req.body="")
    GET /config/v1/agents?service.name=elastic-apm-node&service.environment=development HTTP/1.1
    accept: application/json
    user-agent: elasticapm-node/3.12.0 elastic-apm-http-client/9.5.0 node/12.19.1
    host: localhost:8200
    connection: close
    --
    HTTP/1.1 200 OK
    Date: Wed, 24 Feb 2021 01:25:24 GMT
    Connection: close
    Transfer-Encoding: chunked

    {"log_level":
```

and the APM-using app logged:

```
...
APM Server transport error (200): Unexpected APM Server response when polling config:Unexpected end of JSON input
{"log_level":
...
```

The reason I changed the code here to move the `this.emit(...)` out of the `try { ... }` block was because the `.emit()` is synchronous... and if the handler of that "config" event throws, then it would be caught here in the client code, which isn't desired. I.e. this change is unrelated. I could pull it out to separate change if you prefer.

Reply:
Ah -- my old eyes totally missed that `return`.  This change is all good here -- I appreciate the insights on the practical ins and outs of event based programming.